### PR TITLE
Prevent forever running retrieve request loops

### PR DESCRIPTION
**Pull request by gbalint**
This PR limits request forwarding to 20 hops to avoid forever running request loops.

It introduces a new filed in RetrieveRequestMsg as a **protocol breaking** change. Because of that, stream protocol version is bumped.

### CODE REVIEW

**Review by zelig — APPROVED** [review]

File: swarm/storage/netstore.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ type NetStore struct {
 	closeC            chan struct{}
 }
 
+var fetcherTimeout = 2 * time.Minute // timeout to cancel the fetcher even if requests are coming in
```
Comment (zelig) [inline swarm/storage/netstore.go]: do we really need this?

**Review by zelig — APPROVED** [review]

File: swarm/storage/netstore.go (hunk 1 above)
Comment (janos) [reply swarm/storage/netstore.go]: This is an additional safety cancelation, beside the hop counter limit. It is an internal cancelation, as hop counts are submitted from outside and can be arbitrary counts received.

File: swarm/network/fetcher.go [hunk 2]
```diff
@@ -136,18 +139,22 @@ func (f *Fetcher) Offer(ctx context.Context, source *discover.NodeID) {
...
+func (f *Fetcher) Request(ctx context.Context, hopCount uint64) {
 	// First we need to have this select to make sure that we return if context is done
 	select {
 	case <-ctx.Done():
 		return
 	default:
 	}
 
+	if hopCount >= maxHopCount {
+		return
```
[the remark this answers is not in the document]
Comment (janos) [reply swarm/network/fetcher.go:152]: Yes, you are right. I've added a debug log line.

File: swarm/storage/netstore.go (hunk 1 above)
Comment (zelig) [reply swarm/storage/netstore.go]: understood

**Pull request merged**